Add post on improving accessibility of Rails forms (#153)
x-govuk/x-govuk.github.io: app/posts/finishing-touches-for-forms.md

---
title: Finishing touches for forms
date: 2025-01-05
description: Some tips on making sure forms written in Ruby on Rails are accessible
author:
  name: Peter Yates
  url: https://github.com/peteryates
---

> The [GOV.UK formbuilder](https://govuk-form-builder.netlify.app) makes it easy for Ruby on Rails developers to build forms that follow the rules set out in the [GOV.UK design system](https://design-system.service.gov.uk/).

Reviewer: ```suggestion
The [GOV.UK formbuilder](https://govuk-form-builder.netlify.app) makes it easy for Ruby on Rails developers to build forms that follow the rules set out in the [GOV.UK Design System](https://design-system.service.gov.uk/).
```

> It takes care of the markup, provides a nice API for customising the fields to meet the needs of your service and takes care of clearly dispalying any error messages.

Author: Typo in 'displaying'

> It takes care of the markup, provides a nice API for customising the fields to meet the needs of your service and takes care of clearly dispalying any error messages.

Reviewer: Suggestion for avoiding repeated “takes care of”. Plus maybe HTML is clearer than “markup”?

```suggestion
It generates the correct HTML, provides a nice API for customising the fields to meet the needs of your service and takes care of clearly displaying any error messages.
```

There are, however, some suggestions in the design system guidance that go beyond the remit of the form builder.

They are often forgotten, we'll go over some simple ways to solve them here.

## Positioning the error summary at the top of the page

In addition to making it clear that something is wrong, we also need to tell the user exactly what happened and how to fix it.

[The guidance says](https://design-system.service.gov.uk/components/error-summary#where-to-put-the-error-summary):

> Put the error summary at the top of the main container. If your page includes breadcrumbs or a back link, place it below these, but above the `<h1>`.

This causes a few problems.

> The GOV.UK design system includes some complex rules around linking to fields from the error summary. For example, errors on checkbox or radio button fields should link to the first checkbox or radio button.

Reviewer: This bit isn’t explained in the post. Worth adding a separate section to explain `link_errors: true`?

Reviewer 2: Yeah, this post needs three things; two seems too few!

Reviewer: Header image too! 😀

## Prefixing the page title with `Error:`

When a form submission results in a validation failure we need to make it clear to the user that something is wrong.

[The guidance says](https://design-system.service.gov.uk/patterns/validation/#how-to-tell-the-user-about-validation-errors):

> Add ‘Error: ’ to the beginning of the page `<title>` so screen readers read it out as soon as possible

This is a little tricky because we need a some logic that can determine whether there's an error on the page, and we set the `<title>` in the document `<head>` before we've rendered the form in the `<body>`.

We can use `content_for` to our advantage here, too.

First, we set page titles using a [helper method](https://www.rubyguides.com/2020/01/rails-helpers/) and pass in the text and value that represents whether or not there are any errors on the page:

```html
<%
  page_title(
    "How many implements can you juggle with?",
    error: @object.errors.any?
  )
%>
```

If there are errors we can add the 'Error:' prefix:

```ruby

> def page_title(text, error: false)

Reviewer: Any reason not to recommend our existing helper here? https://govuk-components.netlify.app/helpers/title-with-error-prefix/ 😄

Author: 🤦‍♂️

Oh dear, I forgot it was there. I'll amend the post 😅 Thank you!

Reviewer: Haha, doh! I’m not sure it’s yet widely used in DfE repos as it was added after most of them already had their own helper?

  <%= content_for(:top_of_main) { form.govuk_error_summary } %>

  <%= form.govuk_text_field(:how_many_juggling_items) %>
<% end %>
```

And now our error summary is in the right spot.

![Two screenshots side by side showing a form with errors. The left one has the error summary after the h1 and the right one has the error summary first](/assets/posts/finishing-touches-for-forms/error-summary-side-by-side.png)

## Prefixing the page title with 'Error:'

When a form submission results in a validation failure we need to make it clear to the user that something is wrong.

[The guidance says](https://design-system.service.gov.uk/patterns/validation/#how-to-tell-the-user-about-validation-errors):

> Add ‘Error: ’ to the beginning of the page `<title>` so screen readers read it out as soon as possible

This is a little tricky because we need some logic that can determine whether there's an error on the page, and we set the `<title>` in the document `<head>` before we've rendered the form in the `<body>`.

We can use `content_for` to our advantage here too to place the title where we need it.

```html
<head>
  <%= tag.title(yield(:page_title)) %>
  <!-- other head content -->
</head>

> ```

Reviewer 2: Now this section has been rewritten to mention the helper from GOV.UK Components, I’m a little unsure how this example relates to the helper below. Are they related? Is this trying to demonstrate the old and new, helper-based approach?

Author: I've reworded this. I think the 'Now we' helps here because it's continuing from the step above.

> Now we can the components library's [title with error prefix](https://govuk-components.netlify.app/helpers/title-with-error-prefix/) helper to add the 'Error:' prefix whenever `@object.errors.any?` is `true`.

Reviewer 2: ```suggestion
We can use the [GOV.UK Components](https://govuk-components.netlify.app/) [title with error prefix helper](https://govuk-components.netlify.app/helpers/title-with-error-prefix/) to add the 'Error:' prefix whenever `@object.errors.any?` is `true`.
```

```ruby
<%
  content_for(:page_title) do
    title_with_error_prefix(
      "How many implements can you juggle with?",
      error: @object.errors.any?
    )
  end
%>
```

> ## Making sure our check boxes and radio button error links work

Reviewer: Suggestion for shortening this. I think "checkboxes" is usually 1 word too?

```suggestion
## Making sure error links to checkboxes and radios work
```

You just pass in the list of options and the form builder will render them. This works for simple lists.

Sometimes, however, the list needs to be customised a little further. For example we might need [a conditionally revealed questions](https://design-system.service.gov.uk/components/checkboxes#conditionally-revealing-a-related-question) or [a divider](https://design-system.service.gov.uk/components/checkboxes/#add-an-option-for-none).

To support this the form builder comes with the more powerful `#govuk_check_boxes_fieldset` and `#govuk_radio_buttons_fieldset` methods which let the developer build the form field by field.

This power comes at a cost. For example you could write something like this. Here the `delete` will only be shown to admins:

```ruby
f.govuk_radio_buttons_fieldset(:close_ticket)) do

  if @user.admin?
    f.govuk_radio_button(
      :close_ticket,
      'delete'
    )
  end

  f.govuk_radio_button(
    :close_ticket,
    'archive'
  )

  # the rest of the options
end
```

> This would make it impossible for the error summary to accurately link to the first check box or radio button without repeating the logic --- an approach that's going to lead to bugs when it's updated in one place but not the other.

Reviewer 2: How does `---` render?

Author: Em dash.

> Instead, we have to help the error summary by 'marking' the field we want the error summary to link to with `link_error: true`. This overrides the ID generation so the link in the error summary will match it.

Reviewer: Typo here:

```suggestion
Instead, we have to help the error summary by 'marking' the field we want the error summary to link to with `link_errors: true`. This overrides the ID generation so the link in the error summary will match it.
```

```ruby
f.govuk_radio_buttons_fieldset(:close_ticket)) do

  if @user.admin?
    f.govuk_radio_button(
      :close_ticket,
      'delete',
      link_errors: true
    )
  end

  f.govuk_radio_button(
    :close_ticket,
    'archive',
    link_errors: @user.regular_user?
  )

  # the rest of the options
end
```

> Now when there's a validation error, regardless of whether the user is an admin or not the error summary will link to the first option. It's a good idea to write some tests to ensure future changes don't affect it.

Reviewer 2: Does this need a concluding paragraph? Something about how these finishing touches will make the world of difference to users or something (trite?) like that?

Author: The problem is we'd need a heading to make it clear the summary is for the post rather than for the third tip.

Reviewer: How about a heading of "Your feedback" along the lines of:

> These finishing touches will help make sure your forms are usable and accessible.
>
> If you have any feedback or suggestions of other form features commonly missed, let us know.

Reviewer 2: Alternative is to use an horizontal rule instead of a heading.
